geo/resolver.go

```diff
@@ -0,0 +1,198 @@
+package geo
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"googlemaps.github.io/maps"
+
+	"github.com/bitmark-inc/autonomy-api/schema"
+	"github.com/bitmark-inc/autonomy-api/utils"
+)
+
+var (
+	ErrNoGeoInfoFound         = fmt.Errorf("no geo information found")
+	ErrResolverNotInitialized = fmt.Errorf("location resolver is not initialized")
+)
+
+var (
+	Taiwan = "Taiwan"
+	US     = "United States"
+)
+
+// LocationResolver - interface for resolving location
+type LocationResolver interface {
+	GetPoliticalInfo(schema.Location) (schema.Location, error)
+}
+
+var defaultResolver LocationResolver
+
+type MultipleResolverErrors struct {
+	errors []error
+}
+
+func (e *MultipleResolverErrors) Error() string {
+	errorStrings := make([]string, len(e.errors))
+	for i, err := range e.errors {
+		errorStrings[i] = fmt.Sprintf("#%d: %s", i, err.Error())
+	}
+	return strings.Join(errorStrings, "\n")
+}
+
+func NewMultipleResolverErrors(errors []error) *MultipleResolverErrors {
+	return &MultipleResolverErrors{
+		errors: errors,
+	}
+}
+
+type GeocodingLocationResolver struct {
+	client *maps.Client
+}
+
+func NewGeocodingLocationResolver(client *maps.Client) *GeocodingLocationResolver {
+	return &GeocodingLocationResolver{
+		client: client,
+	}
+}
+
+func (g *GeocodingLocationResolver) GetPoliticalInfo(loc schema.Location) (schema.Location, error) {
+	if loc.Country != "" {
+		return loc, nil
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	geos, err := g.client.Geocode(ctx, &maps.GeocodingRequest{
+		LatLng: &maps.LatLng{
+			Lat: loc.Latitude,
+			Lng: loc.Longitude,
+		},
+		ResultType: []string{"administrative_area_level_2|administrative_area_level_1"},
+		Language:   "en",
+	})
+	if nil != err {
+		return loc, err
+	}
+
+	if len(geos) == 0 {
+		return loc, ErrNoGeoInfoFound
+	}
+
+	var level1, level2 string
+	for _, a := range geos[0].AddressComponents {
+		if len(a.Types) > 0 {
+			switch a.Types[0] {
+			case "administrative_area_level_1":
+				level1 = a.LongName
+			case "administrative_area_level_2":
+				level2 = a.LongName
+			case "country":
+				loc.Country = a.LongName
+			}
+		}
+	}
+
+	loc.Address = geos[0].FormattedAddress
+	loc.County = level2
+
+	switch loc.Country {
+	case US:
+		loc.State = level1
+	case Taiwan:
+		if loc.County == "" {
+			loc.County = utils.EnNameToKey(level1)
+		} else {
+			loc.County = utils.EnNameToKey(loc.County)
+		}
+	default:
```

Comment:
I think we do not need to highlight Iceland here. Because there is not special updates for Iceland. After update, it will become
```
	case US:
	default:
		if loc.County == "" {
			loc.County = level1
		}
	}
```

Reply:
You may be right but I am not sure about default. I would prefer it is an empty string instead of misleading info. If I do not find the info, I can use google geo API to get it or adding for a special case here

Reply:
I do not understand. This package is using google geo api. Could you give me an example how should I change it properly?